ws/adapter.go

```diff
@@ -25,17 +26,44 @@ type Service interface {
 	mainflux.MessagePublisher
 
 	// Subscribes to channel with specified id.
-	Subscribe(uint64, Channel) error
+	Subscribe(uint64, *Channel) error
 }
 
 // Channel is used for receiving and sending messages.
 type Channel struct {
 	Messages chan mainflux.RawMessage
 	Closed   chan bool
+	closed   bool
+	mutex    sync.Mutex
+}
+
+// NewChannel instantiates empty channel.
+func NewChannel() *Channel {
+	return &Channel{
+		Messages: make(chan mainflux.RawMessage),
+		Closed:   make(chan bool),
+		closed:   false,
+		mutex:    sync.Mutex{},
+	}
+}
+
+// Send method send message over Messages channel.
+func (channel *Channel) Send(msg mainflux.RawMessage) {
+	channel.mutex.Lock()
+	defer channel.mutex.Unlock()
```

Comment:
Do we really need `defer`? Explicit Unlock is faster.

Reply:
I am aware that it is faster, but it is more readable this way. I would keep this for now, and we can optimize it later if that kind of optimization is needed.